### PR TITLE
[SPARK-16135][SQL] Remove hashCode and euqals in ArrayBasedMapData

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?

This pr is to remove `hashCode` and `equals` in `ArrayBasedMapData` because the type cannot be used as join keys, grouping keys, or in equality tests.
## How was this patch tested?

Add a new test suite `MapDataSuite` for comparison tests.

### CODE REVIEW

**Comment by srowen** [comment]
LGTM

**Comment by hvanhovell** [comment]
@maropu I think you also need to add these methods to `UnsafeArray` for this to work.

Where in the spark code base do we compare two (unsafe) `MapData` objects? Or are you comparing these in your own code?

**Comment by hvanhovell** [comment]
You could also just also use the approach taken in `UnsafeRow`.

**Comment by maropu** [comment]
It seems `UnsafeArrayData` already has its own `equals` and hashCode`.
Currently, spark doesn't compare unsafe MapData though, I think this might cause implicit bugs.

**Comment by maropu** [comment]
aha, yes. It'd better to take the same approach in `UnsafeRow`?

**Comment by hvanhovell** [comment]
Yeah you are right about `UnsafeArrayData` (my bad).

I would take the same approach as `UnsafeRow`.

**Comment by maropu** [comment]
okay, I'm fixing now.

**Comment by maropu** [comment]
okay, done.

**Comment by mengxr** [comment]
Do we need to hash all values? This could be a performance issue if `hashCode` is called frequently on very large arrays.

Story: MLlib had some performance issues caused by `Vector.hashCode`, which is called during Pyrolite serialization. It saves the Vector as the key in a hash map to avoid re-serialization of the same object. But the `hashCode` costs almost the same as re-serialization.

**Comment by maropu** [comment]
Does the current implementation of `Vector.hashCode` have enough performance? If so, it's okay to follow the impl. to me.

**Comment by hvanhovell** [comment]
The performance of `hashCode()` should be pretty good in this case, and this implementation is inline with the ones used in all other `Unsafe*` objects (MurMurHash). I'd rather be consistent. If this turns out to be a problem, we could always use the first n bytes (similar to the `Vector.hashCode`) for hashCode(), if it turns out to be a problem.

**Comment by maropu** [comment]
At least, we'd be better to leave comments for that.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java [hunk 1]
```diff
@@ -298,6 +298,10 @@ public UnsafeMapData getMap(int ordinal) {
     return map;
   }
 
+  // This `hashCode` computation could consume much processor time for large data.
+  // If the computation becomes a bottleneck, we can use a light-weight logic; the first fixed bytes
+  // are used to compute `hashCode` (See `Vector.hashCode`).
+  // The same issue exists in `UnsafeMapData.hashCode`.
```
Comment (hvanhovell) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java]: The same issue also exists for UnsafeRow...

Comment (maropu) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java]: okay, I'll add now.

**Comment by cloud-fan** [comment]
I think we don't need to implement `equals` and `hashCode` for map type, as map type doesn't support equality and ordering by design, see https://issues.apache.org/jira/browse/SPARK-9415

We should remove the `equals` and `hashCode` on `ArrayBasedMapData` and fix tests if needed

**Comment by maropu** [comment]
Thx, good direction. The current master doesn't throw any exception in an analyzer when map-typed data are passed into `collect_set`/`collect_list`. Probably, should we check the case in there?
https://github.com/apache/spark/pull/13802/files#diff-4b06b5fe0cedf425de14469d1356d6ecR474

**Comment by cloud-fan** [comment]
yea we should improve the type check of `CollectSet`

**Comment by maropu** [comment]
I'm now checking failed tests...

**Comment by maropu** [comment]
@hvanhovell please check again?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala [hunk 2]
```diff
@@ -115,7 +115,7 @@ class CodeGenerationSuite extends SparkFunSuite with ExpressionEvalHelper {
       new GenericArrayData(0 until length),
       new GenericArrayData(Seq.fill(length)(true))))
 
-    if (!checkResult(actual, expected)) {
+    if (!actual.zip(expected).forall { case (data, answer) => checkResult(data, answer)}) {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala]: can we turn map data in `actual` to scala map and compare it with `expected`? (also use scala map in `expected`)

**Comment by cloud-fan** [comment]
looks pretty good, thanks for working on it!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala (hunk 2 above)
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala]: okay, I'll try to replace it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/MapData.scala [hunk 3]
```diff
@@ -39,4 +39,7 @@ abstract class MapData extends Serializable {
       i += 1
     }
   }
+
+  // `MapData` should not implement `equals` and `hashCode` because the type cannot be used as join
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/MapData.scala]: Shall we put this in the class header?

**Comment by hvanhovell** [comment]
LGTM - @cloud-fan?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/MapData.scala (hunk 3 above)
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/MapData.scala]: okay

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/MapData.scala [hunk 4]
```diff
@@ -19,6 +19,10 @@ package org.apache.spark.sql.catalyst.util
 
 import org.apache.spark.sql.types.DataType
 
+/**
+ * `MapData` should not implement `equals` and `hashCode` because the type cannot be used as join
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/MapData.scala]: it's not your fault, but I think we need to add some comment for `MapData` itself, and then follows this comment as a note. We can simply say: `An internal data representation for map type in Spark SQL`.

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/MapData.scala]: okay, how about this?

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM,  retest this as the last test pass is 2 days ago.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.0!

**Comment by cloud-fan** [comment]
thanks, merging to master/2.0!

**Comment by cloud-fan** [comment]
thanks, merging to master/2.0!

**Pull request closed without merging**